packages/enzyme-adapter-utils/src/createMountWrapper.jsx

```diff
@@ -79,17 +103,27 @@ export default function createMountWrapper(node, options = {}) {
       return inst;
     }
 
+    setProviders(providers) {
+      const { providers: currentProviders } = this.state;
+      const providerTypes = providers.map(provider => provider.type);
+
+      this.setState({
+        providers: currentProviders.filter(provider => (
+          !providerTypes.includes(provider.type)
```

Comment:
this API means there's never any way to remove a provider, only to replace existing ones or add new ones.

Separately, `providers` being an array implies that the order matters - why and how? If it matters because it's nested into a tree, then it seems much simpler to only ever allow a *single* provider element, and come up with an alternative solution that lets the user determine their own nesting yet also indicate where children should be rendered.

Reply:
Ah got it! I was a bit confused by what you meant by "a single provider". I was thinking about the new context API in terms of individual values (as is the case with the old API), but perhaps that was me trying to make the new API into something that it is not.

The way you compose multiple contexts in the new React API is by nesting components—perhaps the enzyme API should reflect that, as you've pointed out! So, what do you think about this:

```jsx
mount(<MyComponent />, {
  providers: (
    <ContextA.Provider value="A">
      <ContextB.Provider value="B">
        <ContextC.Provider value="C" />
      </ContextB.Provider>
    </ContextA.Provier>
  ),
});
```

And then changing via context by

```jsx
wrapper.setProviders(
  <ContextA.Provider value="A">
    <ContextB.Provider value="foo">
      <ContextC.Provider value="C" />
    </ContextB.Provider>
  </ContextA.Provier>
);
```

My only concern is that this API makes it a bit challenging to change just a _single_ provider's value without having to re-specify the entire tree. I think it will be common to be dealing with a bunch of different contexts from different libraries, all of which are required for a component to render. But each individual _spec_ will deal with the how the component responds to just _one_ of those contexts changing. So, what do you think about (in addition to the `.setProviders()` API) having an API like this:

```jsx
wrapper.updateProvider(<ContextB.Provider value="foo" />);
```

Which would update _just_ that provider, wherever it is in the tree, and throw an error if the provider is not already in the tree.

Also, I don't feel strongly about any of these names if you'd like them to be different. I am leaning towards keeping the plural "providers" because I think it makes it a bit more clear that the element can have many nested providers, though I do realize you are only passing a single react element.

Reply:
Re-specifying the entire tree is already the case if you want to change a child of the element you wrapped, for example - i'm also not clear on the use case of changing providers after initially wrapping.

Maybe instead of this feature being specific to `createContext` providers, it could be more general - something like, "wrapWith". A single render prop function that gets passed a child, and wraps it as needed.

Example:
```jsx
const wrapper = mount(<Foo />, {
  wrapWith(root) {
    return (
      <ContextA.Provider value="A">
        <ContextB.Provider value="foo">
          <ContextC.Provider value="C">
            {root}
          </ContextC.Provider>
        </ContextB.Provider>
      </ContextA.Provider>
    );
  },
});
```
The API could throw if your wrapWith function failed to render the child (solo, ie, with no siblings), and then `wrapper` in this case would be `<Foo />`, not any of the providers, but they'd still be exercised in the rendering path.

Reply:
I think the use-case is something like this:

Imagine you have a [`react-flag`](https://github.com/garbles/flag)-like library and some future version of `react-redux` that uses the new context API. You'd render your component like this (using the wrapWith API):

```jsx
// my app's feature flags
const flags = {
  foo: false,
  bar: false,
};

const wrapper = mount(<Foo />, {
  wrapWith: root => (
    <FlagProvider value={flags}>
      <ReduxProvider value={store}>{root}</ReduxProvider>
    </FlagProvider>
  ),
});
```

And now we want to test that some side-effect in `<Foo />`'s `componentDidUpdate` gets called when the feature flag changes. We need some way of changing `<FlagProvider />`'s `value`, and it would be nice if we didn't have to re-setup the `<ReduxProvider />` when we do that.

Also, if our component starts depending on a new provider in the future, it'd be really nice to only have to add it to the options we pass enzyme, and not every place where we're just changing a single provider's value.

I'm guessing the API for changing the `wrapWith` function could be something like

```jsx
wrapper.setWrapWith(root => (
  <FlagProvider value={{ ...flags, bar: true }}>
    <ReduxProvider value={store}>{root}</ReduxProvider>
  </FlagProvider>
));
```

But that doesn't address my concern about having to respecify the entire tree. One solution that comes to mind (just spitballing here) is making `wrapWith` a _component_:

```jsx
const wrapper = mount(<Foo />, {
  wrapper: {
    component: (props) => {
      const { flags, store, children } = props;

      return (
        <FlagProvider value={{ ...flags, bar: true }}>
          <ReduxProvider value={store}>{children}</ReduxProvider>
        </FlagProvider>
      );
    },
    props: { store: createStore, flags },
  },
});
```

And then having an API that lets you set the wrapper component's props:

```jsx
wrapper.setWrapperProps({
  flags: {
    ...flags,
    foo: true,
  },
});
```

(`wrapWith` seemed like a better name for a function than a component so I changed it to just `wrapper`; but it also seems a bit confusing because `wrapper` is the conventional name for a `ReactWrapper` instance.)

Reply:
(We can bikeshed the names later)

OK, so you're suggesting the user be able to provide a "wrapping component", and provide explicit `setProps`/`setState` access on it - what about instead, `wrapper.getWrappingComponent()`, that returned a normal enzyme wrapper around that component - providing you with the same APIs?

Reply:
I like that idea a lot! I'll implement and then drop you a comment here so we can discuss. I really appreciate the time you've taken to work through this with me. Thank you!